File: NameKeeper/app/src/main/java/pe/edu/upc/namekeeper/MainActivity.java
package pe.edu.upc.namekeeper;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {
    TextView nameTextView;
    // Preferences parameters
    private static final String PREFS = "prefs";
    private static final String PREF_NAME = "name";
    // Preferences object
    SharedPreferences sharedPreferences;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        nameTextView = (TextView) findViewById(R.id.nameTextView);
        updateCurrentName("");

        FloatingActionButton fab =
                (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                setKeptName("");
                Snackbar.make(view, "Name has been cleared",
                        Snackbar.LENGTH_LONG)
                        .setAction("Keep Name", new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                displayWelcome();
                            }
                        }).show();
                updateCurrentName("");
            }
        });
        displayWelcome();
    }

    private void displayWelcome() {
        // Presents an Alert Dialog to request a name if not present
        String name = getKeptName();
        if(name.length() > 0) {
            // A name was found, so it welcomes back the user
            Toast.makeText(this, "Welcome back " + name + "!",
                    Toast.LENGTH_LONG).show();
            updateCurrentName(name);
        }
        else {
            // No name is present, so it needs to request one
            AlertDialog.Builder alert = new AlertDialog.Builder(this);
            alert.setTitle("Name Keeper");
            alert.setMessage("What is your name?");
            final EditText nameEditText = new EditText(this);
            alert.setView(nameEditText);
            alert.setPositiveButton("OK",
                    new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            String name = nameEditText.getText().toString();
                            if (name.length() > 0) {
                                setKeptName(name);
                                Toast.makeText(getApplicationContext(),
                                        "Welcome " + name + "!",
                                        Toast.LENGTH_LONG).show();
                                updateCurrentName(name);
                            }
                        }
                    });
            alert.setNegativeButton("Cancel",
                    new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            Toast.makeText(getApplicationContext(),
                                    "Don't worry, no pressure",
                                    Toast.LENGTH_LONG).show();
                        }
                    });
            alert.show();
        }
    }

    private void setKeptName(String name) {
        // Stores given name in Preferences
        if(sharedPreferences != null) {
            // Opens Preferences file for edition
            SharedPreferences.Editor e = sharedPreferences.edit();
            e.putString(PREF_NAME, name);
            e.commit();
        }
    }
    private String getKeptName() {
        // Obtains stored name if present,
        // otherwise returns an empty string
        if(sharedPreferences == null) {
            sharedPreferences = getSharedPreferences(
                    PREFS, MODE_PRIVATE);
        }
        return sharedPreferences.getString(PREF_NAME, "");
    }

    private void updateCurrentName(String name) {
        // Update TextView depending on name
        String nameMessage = name.length() > 0 ?
                "Current name is " + name :
                "No name kept";
        nameTextView.setText(nameMessage);
    }
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: ormdemo/app/src/main/java/pe/com/ormdemo/MainActivity.kt
package pe.com.ormdemo

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.Contacts
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import com.google.android.material.snackbar.Snackbar
import com.orm.SugarRecord.*
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    lateinit var nameEditText: EditText
    lateinit var emailEditText: EditText
    lateinit var genderSwitch: Switch
    lateinit var malesProgressBar: ProgressBar
    lateinit var femalesProgressBar: ProgressBar
    lateinit var malesMaxTextView: TextView
    lateinit var femalesMaxTextView: TextView
    lateinit var male: Gender
    lateinit var female: Gender

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        nameEditText = findViewById(R.id.nameEditText)
        emailEditText = findViewById(R.id.emailEditText)
        genderSwitch = findViewById(R.id.genderSwitch)
        malesProgressBar = findViewById(R.id.malesProgressBar)
        femalesProgressBar = findViewById(R.id.femalesProgressBar)
        malesMaxTextView = findViewById(R.id.malesMaxTextView)
        femalesMaxTextView = findViewById(R.id.femalesMaxTextView)
        initDb()
        fab.setOnClickListener { view ->
            val name = nameEditText.text.toString()
            val email = emailEditText.text.toString()
            val gender = if (genderSwitch.isChecked) female else male
            var resultMessage = "Person added"
            if (name.isEmpty()) {
                resultMessage = "Name is empty"
            } else {
                val people = find(Person::class.java, "name = ?", name)
                if (people.isEmpty()) {
                    var person = Person(name, email, gender)
                    person.save()
                    val id = person.id
                    person = findById(
                        Person::class.java, id
                    )
                    Log.d(
                        "ORMDemo",
                        "Person id: " + person.id.toString() +
                                ", email: " + person.email +
                                ", gender: " + person.gender.name
                    )
                    updateProgress()
                } else {
                    resultMessage = "Person name already exists"
                }
            }

            Snackbar.make(
                view,
                resultMessage,
                Snackbar.LENGTH_LONG
            )
                .setAction("New Person") {
                    nameEditText.setText("")
                    emailEditText.setText("")
                    genderSwitch.isChecked = false
                }.show()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId


        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)

    }

    private fun initDb() {
        val genders = listAll(Gender::class.java)
        val people = listAll(Person::class.java)
        if (genders.isEmpty()) {
            male = Gender("Male")
            female = Gender("Female")
            male.save()
            female.save()

        } else {
            male = find(Gender::class.java, "name = ?", "Male").get(0)
            female = find(Gender::class.java, "name = ?", "Female").get(0)
        }
        if (!people.isEmpty()) {
            Log.i("PEOPLE", people.toString())
        }
        updateProgress()
    }

    fun clearAll(view: View) {
        deleteAll(Person::class.java)
        deleteAll(Gender::class.java)
        initDb()
    }

    fun updateProgress() {
        val malesCount = male.getPeople()?.size
        val femalesCount = female.getPeople()?.size
        val peopleCount = listAll(Person::class.java)?.size
        val malesProgress = if (peopleCount == 0) 0 else Math.round((malesCount!! * 100 / peopleCount!!).toFloat())
        val femalesProgress = if (peopleCount == 0) 0 else 100 - malesProgress
        malesProgressBar.setProgress(malesProgress)
        malesMaxTextView.text = "$malesProgress %"
        femalesProgressBar.setProgress(femalesProgress)
        femalesMaxTextView.text = "$femalesProgress %"
    }
}
